### ISSUE TITLE
Performance overhead on browsers without support for registerElement

### ISSUE THREAD

**Issue by wookieb** [skatejs/skatejs#349]
I've noticed a huge performance overhead when inserting big amount of elements to the document on browsers without support for "registerElement".

Example here
http://jsfiddle.net/js63vqaf/4/

In profiler I see that javascript spent almost 1 second just in skate while inserting the content took 300ms.
Is it possible to make it faster? Or split skate's analysis of inserted content into async batches?

**Comment by treshugart** [comment]
I think the biggest overhead is where we're trying to make the native custom element lifecycle have some semblance of order. We're actually talking about removing this before the next release as it's extremely difficult and costly. The spec should have done a better job of making this consistent but they didn't. It'll just affect how components are built, and possibly for the better anyways (less coupling).

The other aspect to performance is that there must be a mutation observer and the mutation observer handler must interrogate each element that was added for information in order for Skate to upgrade it or leave it alone. This part has already been optimised to the hilt. If using Skate's mutation observer, it should be about 10x faster than Polymer's (it's been a few months since I tested). 

I'm going to experiment with removing some of that code that walks the tree of every component to see if it's something we want to do. We should see a perf gain from that. I don't think there's much we can do in the Mutation Observer but maybe there's some optimisations we can do in the created lifecycle where we do a lot of the magic in Skate.

Just for some comparisons to see how we're doing, I created a couple rough benchmarks based on your Fiddle to run in Chrome (Skate's uses ES6 template literals for readability). See: https://gist.github.com/treshugart/996cf181da4fba92164a.

I'm still awaiting feedback on the React one, but it seems to be the simplest I could get it.

**Comment by treshugart** [comment]
Wow. Just removed the bits of code around with (only broke 3 tests around ensuring lifecycle ordering) and cut off a lot of overhead.

- Skate: 4.75s
- Skate native: 1.75s

Issue for removing: #351.

**Comment by wookieb** [comment]
So are saying that React took 9s to render something. Could you provide more detailed informations? Code or smth?

**Comment by treshugart** [comment]
Yeah, the link to the gist is in my original comment. It contains the code of the slower Skate code, though, but I'll update once we finalise whatever is happening there.

**Comment by treshugart** [comment]
https://gist.github.com/treshugart/996cf181da4fba92164a for convenience.

**Comment by treshugart** [comment]
I'll add that I had @alexreardon peer review the React example to make sure I wasn't doing anything stupid. The original goal was never to pit them together but to be able to draw a baseline from an alternative that is considered "fast" to make sure we're within an acceptable performance level for this particular use case.

**Comment by treshugart** [comment]
Some performance enhancements around this were implemented in #352. Instead of forcefully initialising components, there is now an opt-in API method for executing a callback on a set of elements when they're all ready:

```js
skate.ready(elements, function (elements) {
   // elements are ready
});
```
